trie/iterator.go

```diff
@@ -836,3 +836,41 @@ func (it *unionIterator) Error() error {
 	}
 	return nil
 }
+
+// prefixIterator is a wrapper around NodeIterator that stops iteration
+// when it leaves a subtree with a specific prefix.
```

Comment:
I would prefer just adding extra fields to the current iterator: this is to be used outside of the normal geth operations, and if we use a different object, it will mean extra maintenance for something we don't use.

Add all the new fields you need to the regular iterator, make `NewIterator` leave them empty and only fill them inside `NewSubtreeIterator`.

Reply:
I initially went with a wrapper to avoid touching the core iterator - seemed safer at the time. But you're spot on, it's unnecessary complexity. Adding the fields directly is way cleaner.